server/src/shared/app/typed-controller.ts

```diff
@@ -0,0 +1,65 @@
+import { z } from 'zod'
+import express from 'express'
+import { Result, Ok, Err } from '../core/result'
+import { ValidationError } from '../core/validation'
+import { BaseController } from './base-controller'
+import { UseCaseWithDTO } from './use-case-with-dto'
+
+type getArgs<UseCase> = UseCase extends UseCaseWithDTO<infer T, any> ? T : never
+type getResult<UseCase> = UseCase extends UseCaseWithDTO<any, infer T> ? T : never
+
+export abstract class TypedController<
```

Comment:
I know we spent a long time on the implementation of the Controller during Winter '22. You've spent more time with the codebase since then so if you feel like moving to one unified controller with empty objects for GET requests then I support the change! Maybe get @KTong821's opinion when he's back?

Reply:
I needed a controller that used `zod` instead of Joi, so I built this controller to do that. Scope creep added a bunch of other possible upgrades. This change is an open request for feedback that extends beyond this PR. The old controllers will stick around until we decide what to do with them.

My two cents:
I think one of the reasons we wanted `withoutDTO` stuff is that we didn't think we needed validation for endpoints that don't have arguments. Thinking about it more, we should actually validate that the request body / query params are empty. Then, a type-validated controller that produces empty arguments is perhaps still appropriate for argless use cases.